--- REL_17_5_WASM/src/backend/libpq/pqcomm.c
+++ pglite-REL_17_5_WASM/src/backend/libpq/pqcomm.c
@@ -131,6 +131,8 @@
 static char *PqRecvBuffer;
 static int	PqRecvPointer;
 static int	PqRecvLength;
+volatile int querylen = 0;
+volatile FILE* queryfp = NULL;
 #endif
 
 /*
@@ -320,8 +322,8 @@
 	AddWaitEventToSet(FeBeWaitSet, WL_POSTMASTER_DEATH, PGINVALID_SOCKET,
 					  NULL, NULL);
 #else /* WASM */
-    PDEBUG("# 220: FIXME: socketfile");
-    #pragma message "FIXME: socketfile"
+    PDEBUG("# 323: FIXME: socketfile");
+    #pragma message "FIXME: use socketfile when overflowing PqRecvBuffer_static"
     /* because we fill before starting reading message */
     PqRecvBuffer = &PqRecvBuffer_static[0];
 #endif /* WASM */
@@ -923,6 +925,20 @@
 		else
 			PqRecvLength = PqRecvPointer = 0;
 	}
+#if defined(__EMSCRIPTEN__) || defined(__wasi__)
+    if (queryfp && querylen) {
+        int got = fread( PqRecvBuffer, 1, PQ_RECV_BUFFER_SIZE - PqRecvPointer, queryfp);
+        querylen -= got;
+        PqRecvLength += got;
+        if (querylen<=0) {
+            PDEBUG("# 931: could close fp early here " __FILE__);
+            queryfp = NULL;
+        }
+        if (got>0)
+    		return 0;
+    }
+    return EOF;
+#endif
 
 	/* Ensure that we're in blocking mode */
 	socket_set_nonblocking(false);
@@ -1025,7 +1041,9 @@
 		*c = PqRecvBuffer[PqRecvPointer++];
 		return 1;
 	}
-
+#if defined(__EMSCRIPTEN__) || (__wasi__)
+puts("# 1044: pq_getbyte_if_available N/I in " __FILE__ ); abort();
+#else
 	/* Put the socket into non-blocking mode */
 	socket_set_nonblocking(true);
 
@@ -1062,7 +1080,7 @@
 		/* EOF detected */
 		r = EOF;
 	}
-
+#endif
 	return r;
 }
 
@@ -1129,6 +1147,7 @@
 	return 0;
 }
 
+
 /* --------------------------------
  *		pq_buffer_remaining_data	- return number of bytes in receive buffer
  *
@@ -1150,16 +1169,22 @@
  *		This must be called before any of the pq_get* functions.
  * --------------------------------
  */
-#if defined(I_EMSCRIPTEN) || defined(I_WASI)
+#if defined(__EMSCRIPTEN__) || defined(__wasi__)
 EMSCRIPTEN_KEEPALIVE void
 pq_recvbuf_fill(FILE* fp, int packetlen) {
-    fread( PqRecvBuffer, packetlen, 1, fp);
+    if (packetlen>PQ_RECV_BUFFER_SIZE) {
+        int got = fread( PqRecvBuffer, 1, PQ_RECV_BUFFER_SIZE, fp);
+        queryfp = fp;
+        querylen = packetlen - got;
+        PqRecvLength = got;
+PDEBUG("# 1178: input overflow");
+    } else {
+        fread( PqRecvBuffer, packetlen, 1, fp);
+        PqRecvLength = packetlen;
+        queryfp = NULL;
+        querylen = 0;
+    }
     PqRecvPointer = 0;
-    PqRecvLength = packetlen;
-#if PDEBUG
-        printf("# 1199: pq_recvbuf_fill cma_rsize=%d PqRecvLength=%d buf=%p reply=%p\n", cma_rsize, PqRecvLength, &PqRecvBuffer[0], &PqSendBuffer[0]);
-#endif
-
 }
 #endif
 extern int cma_rsize;
@@ -1175,7 +1200,7 @@
 		ereport(FATAL,
 				(errcode(ERRCODE_PROTOCOL_VIOLATION),
 				 errmsg("terminating connection because protocol synchronization was lost")));
-#if defined(I_EMSCRIPTEN) || defined(I_WASI)
+#if defined(__EMSCRIPTEN__) || defined(__wasi__)
     if (!pq_buffer_remaining_data()) {
         if (cma_rsize) {
             PqRecvPointer = 0;
@@ -1326,7 +1351,7 @@
 static int
 internal_putbytes(const char *s, size_t len) {
 	if (PqSendPointer >= PqSendBufferSize) {
-        fprintf(stderr, "# 1329: overflow %d >= %d cma_rsize=%d CMA=%d\n", PqSendPointer, PqSendBufferSize,cma_rsize, CMA_MB);
+        fprintf(stderr, "# 1329: overflow %zu >= %d cma_rsize=%d CMA=%d\n", PqSendPointer, PqSendBufferSize,cma_rsize, CMA_MB);
     }
 
     if (!cma_rsize) {
@@ -2179,4 +2204,4 @@
 	}
 
 	return true;
-}
+}
\ No newline at end of file
